# Assignment: Predict user conversion based on session data

**Dataset:** Online Shoppers Purchasing Intention Dataset
<br>
**Credits:** Sakar, C.O., Polat, S.O., Katircioglu, M. et al. Neural Comput & Applic (2018)
<br>
**Link to dataset:** https://www.kaggle.com/datasets/henrysue/online-shoppers-intention/data


### Assignment description

An E-commerce company named KokoBananas that sells health and cosmetic products have realised that, in order to stay competitive, they need a better understanding of what constitutes a "good" user experience (UX) on their site. It is assumed that users with a "good" UX converts - so *conversion* is the key metric in this assignment.

Using historical session data, analyse which user behavior leads to conversion. The goal is to create an AI-model, that will, *in real time*, be able to predict whether a user session will convert. The idea is to apply this model on future users in order to quickly determine which user actually become paying customers.

You will present your findings to the company management team. You need to remain object and present both strengths and weaknesses in your predictions. You might even have to report back that it is not possible to use the available data for this prediction. Be mindful of not being too optmistic about your findings - we often fall into the trap of getting swayed by high accuracy - but we always need to criticise this and understand *why* we are getting high accuracy, and if it actually will lead to useful predictions in the future.

The presentation shall be included in the Jupyter Notebook used for analysing the data and training models. The presentation should be understandable for a non-technical person and contain actionable insights. The company management should be helped in making correct decisions based on objective and easy-to-understand results.

### Some initial thoughts on the assignment

- Fun and useful real world application!
- Very open - we are free to use any model we have come across during the course - experimentation nation!
- Will be a good vechile of rehearsing the full course content and applying it throughout.
- It can surely be debated whether conversion is a useful metric for "good" UX. You may have a site that is great at conversion but the users may feel that the UX is awful. "Good" UX can be a very subjective and emotional experience.
- It is a bit vague what the findings will in practice be used for. There is a desire to know in real time which user will convert and at a later stage improve the UX? Will the UX be changed to optimise for conversion? Will Users who are about to drop out of a session without converting be nudged in some way?

### General guidance for the assignment

Remember to keep the code simple and easy to follow. Make use of proper comments, docstrings, typehints etc.
Use markdown cells to, in the simplest terms, motivate your choices.

Since the ambition is to, in real time, be able to identify which users will convert, it is important that the inference time is rapid. So the model we use needs to be quick. Depending on the results of experimenting with models and the data, one outcome might be that it's difficult to get a high performance model that also has good predictions.

This is a binary classification exercise. Use this strategy:
- EDA
- Feature Engineering
- Rehearse all classification models used in the course
- Hypothesize which one/ones might be best
- Try all classification models we have used during the course
- See if your hypothesis held or not with the results
- Compare with the paper written by the authors of the dataset

In [111]:
# packages and methods to import

import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
from IPython.display import display

from sklearn.impute import KNNImputer

In [112]:
# Read the csv file of session data.
project_data_path =  'project_data.csv'
original_data_path = 'online_shoppers_intention_original_dataset.csv'

session_data_df = pd.read_csv(project_data_path)
original_data_df = pd.read_csv(original_data_path)

session_data_df.info()

session_data_df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12330 entries, 0 to 12329
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Unnamed: 0               12330 non-null  int64  
 1   Administrative           12330 non-null  int64  
 2   Administrative_Duration  12330 non-null  float64
 3   Informational            12330 non-null  int64  
 4   Informational_Duration   12330 non-null  float64
 5   ProductRelated           12330 non-null  int64  
 6   ProductRelated_Duration  12330 non-null  float64
 7   BounceRates              12330 non-null  float64
 8   ExitRates                12330 non-null  float64
 9   PageValues               12330 non-null  float64
 10  SpecialDay               12207 non-null  float64
 11  Month                    12330 non-null  object 
 12  OperatingSystems         12330 non-null  int64  
 13  Browser                  12146 non-null  float64
 14  Region                

,Unnamed: 0,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration,BounceRates,ExitRates,PageValues,SpecialDay,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
0,0,0,0.0,0,0.0,1,0.000000,0.200000,0.200000,0.000000,0.0,Feb,1,NaN,1.0,1,Returning_Visitor,False,False
1,1,0,0.0,0,0.0,2,64.000000,0.000000,0.100000,0.000000,0.0,Feb,2,2.0,NaN,2,Returning_Visitor,False,False
2,2,0,0.0,0,0.0,1,0.000000,0.200000,0.200000,0.000000,0.0,Feb,4,1.0,NaN,3,Returning_Visitor,False,False
3,3,0,0.0,0,0.0,2,2.666667,0.050000,0.140000,0.000000,0.0,Feb,3,2.0,2.0,4,Returning_Visitor,False,False
4,4,0,0.0,0,0.0,10,627.500000,0.020000,0.050000,0.000000,0.0,Feb,3,3.0,1.0,4,Returning_Visitor,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12325,12325,3,145.0,0,0.0,53,1783.791667,0.007143,0.029031,12.241717,0.0,Dec,4,6.0,NaN,1,Returning_Visitor,True,False
12326,12326,0,0.0,0,0.0,5,465.750000,0.000000,0.021333,0.000000,0.0,Nov,3,2.0,1.0,8,Returning_Visitor,True,False
12327,12327,0,0.0,0,0.0,6,184.250000,0.083333,0.086667,0.000000,0.0,Nov,3,2.0,1.0,13,Returning_Visitor,True,False
12328,12328,4,75.0,0,0.0,15,346.000000,0.000000,0.021053,0.000000,0.0,Nov,2,2.0,3.0,11,Returning_Visitor,False,False


A quick glance at the basic info() summary of the dataset and looking at the first and last 5 rows of the dataset tells us:
1. First column: "Unnamed: 0" is an unnecessary extra index column created when we read the csv. This can be dropped immediately
2. A few columns have nans - lets put these in a dict for easy access later.

In [113]:
# Drop the extra index column. Lets use inplace=True since we wont need this column later.
session_data_df.drop('Unnamed: 0', axis=1, inplace=True)

Verbose method for storing columns with nans in a dict.

In [116]:
# Find columns with Nans (.isna() method)
# calculate the sum of Nans in each column (.sum() method) 
# store it in a dict (.to_dict() method) so we can loop over it in the next step
nans_per_column = session_data_df.isna().sum().to_dict()

# create an empty dict where we can store the columns that have nans and their respective values.
columns_with_nans = {}

# loop over all key value pairs in our nans per column dict
for key, value in nans_per_column.items():

    # if a column has a non zero amount of nans, add that column and the amount of nans to our columns with nans dict.
    if value != 0:

        columns_with_nans[key] = value

# we can use this dict later in our EDA.
columns_with_nans

{'SpecialDay': 123, 'Browser': 184, 'Region': 246, 'Revenue': 147}

Using dictionary comprehension, we can use one-line of code to store columns with nans in a dict.

In [117]:
# Use .isna() together with .sum() to find amount of nans in each column
# convert that into a dict with .to_dict()
# use the .items() method so we can loop over the key, value pairs and check where the value is non-zero.
# key value pairs with non-zero values are stored in a columns with nans dict!
columns_with_nans = {key : value for key, value in session_data_df.isna().sum().to_dict().items() if value != 0}

In [118]:
columns_with_nans

{'SpecialDay': 123, 'Browser': 184, 'Region': 246, 'Revenue': 147}

# REUSEABLE ITEMS

Throughout the analysis we need easy access to the descriptions of each feature column, including what the values mean.

We can create a dictionary that contains this information. The descriptions of the features are not that detailed or in some cases, also not clear.

The unit of measurement for time spent within a page category is not given.

Furthermore we can group the feature columns for easier viewing later on. Lets create some lists that will be easy to iterate over.

In [119]:
# Dictionary with all the columns + description of each. The decription in the dict is my interpretation of the source
# documentation which was sparse on column descriptions.
column_descriptions = {
                    "Administrative" : "Number of page visits within this category",
                    "Administrative_Duration" : "Time spent within this category",
                    "Informational" : "Number of page visits within this category",
                    "Informational_Duration" : "Time spent within this category", 
                    "ProductRelated" : "Number of page visits within this category", 
                    "ProductRelated_Duration" : "Time spent within this category",
                    "BounceRates" : "Bounce rate of current page",
                    "ExitRates" : "Exit rate value of current page",
                    "PageValues" : "Page value of current page",
                    "SpecialDay" : "Proximity in time of the session to a special day (e.g. Mother’s Day, Valentine's Day) in which the sessions are more likely to be finalized with transaction. The value of this attribute is determined by considering the dynamics of e-commerce such as the duration between the order date and delivery date. For example, for Valentina’s day, this value takes a nonzero value between February 2 and February 12, zero before and after this date unless it is close to another special day, and its maximum value of 1 on February 8.",
                    "Month" : "Month of session",
                    "Weekend" : "Boolean if the session day was in the weekend",
                    "OperatingSystems" : "Different OS' represented by different integers",
                    "Browser" : "Different browsers represented by different integers",
                    "Region" : "Different regions represented by different integers",
                    "TrafficType" : "Different traffic types represented by different integers",
                    "VisitorType" :  "Different visitor types represented by different categorical values",
                    }

# Lists grouping together columns of similar kind.
column_info_page_views_and_duration_list = ['Administrative', 'Administrative_Duration', 'Informational','Informational_Duration', 'ProductRelated','ProductRelated_Duration']
column_info_google_analytics_list = ['BounceRates','ExitRates','PageValues']
column_info_time_categories_list = ['SpecialDay', 'Month', 'Weekend']
column_info_misc_categories_list = ['OperatingSystems','Browser','Region','TrafficType','VisitorType']

In [120]:
# Code to check that we actually added all columns to our dict
columns_missing_from_description_dict = set(session_data_df.columns) - set(column_descriptions.keys())

print(f"You missed adding these columns to your column description dict: {columns_missing_from_description_dict}. If it's revenue, that's ok since it's our target column.")

You missed adding these columns to your column description dict: {'Revenue'}. If it's revenue, that's ok since it's our target column.


In [121]:
# Code to check that we actually added all columns to our dict
columns_missing_from_grouped_columns_list = set(session_data_df.columns) - set(column_info_misc_categories_list) - set(column_info_time_categories_list) - set(column_info_google_analytics_list) - set(column_info_page_views_and_duration_list)

print(f"You missed adding these columns to your grouped column list: {columns_missing_from_grouped_columns_list}. If it's revenue, that's ok since it's our target column.")

You missed adding these columns to your grouped column list: {'Revenue'}. If it's revenue, that's ok since it's our target column.


In [122]:
session_df_row_count = len(session_data_df)
print(session_df_row_count)

target_column_before_cleaning = session_data_df['Revenue']

target_true_count_before_cleaning = len(session_data_df.loc[session_data_df['Revenue'] == True])
print(target_true_count_before_cleaning)

target_false_count_before_cleaning = len(session_data_df.loc[session_data_df['Revenue'] == False])
print(target_false_count_before_cleaning)

delimiter_between_prints = '\n\n-----------\n\n'

12330
1880
10303


# REUSEABLE FUNCTIONS

In [123]:
# create a function for summarising values in a dataframe

def summarise_df(df: pd.DataFrame,
                column_info_page_views_and_duration_list:list=column_info_page_views_and_duration_list,
                column_info_google_analytics_list:list=column_info_google_analytics_list, 
                column_info_time_categories_list:list=column_info_time_categories_list) -> list:
    
    # Create four empty dicts according the category groups outlined earlier
    # We will in the loops below add summarised values from each column to these dicts.
    column_info_page_views_and_duration_summary = {}
    column_info_google_analytics_summary = {}
    column_info_time_categories_summary = {}
    column_info_misc_categories_summary = {}

    # Iterate over each column in the session DataFrame
    for col in df.columns:

        # The Pandas API types module provides efficient functions for checking datatypes.
        # First lets check if the column contains numerical values.
        if pd.api.types.is_numeric_dtype(df[col]):

            # Next lets check which category grouping it belongs to.
            if col in column_info_page_views_and_duration_list:

                # For numerical regression data we want to know the max, min, mean values and the standard deviation
                column_info_page_views_and_duration_summary[col] = {
                                    'data_type' : df[col].dtype,
                                    'max_value' : df[col].max(),
                                    'min_value' : df[col].min(),
                                    'mean_value' : df[col].mean(),
                                    'standard_deviation' : df[col].std()
                                    }
                
            elif col in column_info_google_analytics_list:

                column_info_google_analytics_summary[col] = {
                                    'data_type' : df[col].dtype,
                                    'max_value' : df[col].max(),
                                    'min_value' : df[col].min(),
                                    'mean_value' : df[col].mean(),
                                    'standard_deviation' : df[col].std()
                                    }
                
            elif col in column_info_time_categories_list:

                column_info_time_categories_summary[col] = {
                                    'data_type' : df[col].dtype,
                                    'max_value' : df[col].max(),
                                    'min_value' : df[col].min(),
                                    'mean_value' : df[col].mean(),
                                    'standard_deviation' : df[col].std()
                                    }
            else:
                # For classification data, numerical or object, we want to know the unique values, their counts (including NaNs!) and top & bottom
                list_of_tuples_of_key_value_pairs_from_value_counts = [(key, int(value)) for key, value in df[col].value_counts(dropna=False).to_dict().items()]
                column_info_misc_categories_summary[col] = {
                                    'data_type' : df[col].dtype,
                                    'unique_values' : df[col].unique(),
                                    'value_counts' : list_of_tuples_of_key_value_pairs_from_value_counts,
                                    'top_value' : list_of_tuples_of_key_value_pairs_from_value_counts[0],
                                    'bottom_value' : list_of_tuples_of_key_value_pairs_from_value_counts[-1]
                                    }
        
        else:
            list_of_tuples_of_key_value_pairs_from_value_counts = [(key, int(value)) for key, value in df[col].value_counts(dropna=False).to_dict().items()]
            column_info_misc_categories_summary[col] = {
                                    'data_type' : df[col].dtype,
                                    'unique_values' : df[col].unique(),
                                    'value_counts' : list_of_tuples_of_key_value_pairs_from_value_counts,
                                    'top_value' : list_of_tuples_of_key_value_pairs_from_value_counts[0],
                                    'bottom_value' : list_of_tuples_of_key_value_pairs_from_value_counts[-1]
                                    }

    # convert summary dicts into dataframes for better visualisation   
    page_views_and_duration_summary_df = pd.DataFrame(column_info_page_views_and_duration_summary)  
    column_info_google_analytics_summary_df = pd.DataFrame(column_info_google_analytics_summary)
    column_info_time_categories_summary_df = pd.DataFrame(column_info_time_categories_summary)
    column_info_misc_categories_summary_df = pd.DataFrame(column_info_misc_categories_summary)

    summary_dfs_list = [page_views_and_duration_summary_df, column_info_google_analytics_summary_df, column_info_time_categories_summary_df, column_info_misc_categories_summary_df]
    
    # display dataframes with HTML in notebooks using the display method
    for df in summary_dfs_list:

        display(df)
    # display(page_views_and_duration_summary_df)
    # display(column_info_google_analytics_summary_df)
    # display(column_info_time_categories_summary_df)
    # display(column_info_misc_categories_summary_df)

    return summary_dfs_list

In [124]:
summary_dfs_list = summarise_df(session_data_df)

,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration
data_type,int64,float64,int64,float64,int64,float64
max_value,27,989493.0,24,2549.375,705,63973.52223
min_value,-10,0.0,0,0.0,0,0.0
mean_value,2.2382,1269.776004,0.503569,34.472398,31.731468,1194.74622
standard_deviation,3.409269,34071.46693,1.270156,140.749294,44.475503,1913.669288


,BounceRates,ExitRates,PageValues
data_type,float64,float64,float64
max_value,2.098952,0.2,361.763742
min_value,-0.49868,0.0,0.0
mean_value,0.044239,0.043073,5.889258
standard_deviation,0.239735,0.048597,18.568437


,SpecialDay
data_type,float64
max_value,1.0
min_value,0.0
mean_value,0.06126
standard_deviation,0.19865


,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
data_type,object,int64,float64,float64,int64,object,object,object
unique_values,"[Feb, Mar, May, Turc, Oct, June, Jul, Aug, Nov...","[1, 2, 4, 3, 7, 6, 8, 5]","[nan, 2.0, 1.0, 3.0, 4.0, 5.0, 6.0, 7.0, 10.0,...","[1.0, nan, 2.0, 3.0, 4.0, 9.0, 5.0, 6.0, 7.0, ...","[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...","[Returning_Visitor, New_Visitor, Other]","[False, True, Name:Zara]","[False, nan, True]"
value_counts,"[(May, 3196), (Nov, 2998), (Mar, 1907), (Dec, ...","[(2, 6601), (1, 2585), (3, 2555), (4, 478), (8...","[(2.0, 7835), (1.0, 2424), (4.0, 725), (5.0, 4...","[(1.0, 4695), (3.0, 2345), (4.0, 1157), (2.0, ...","[(2, 3913), (1, 2451), (3, 2052), (4, 1069), (...","[(Returning_Visitor, 10551), (New_Visitor, 169...","[(False, 9273), (True, 2868), (Name:Zara, 189)]","[(False, 10303), (True, 1880), (nan, 147)]"
top_value,"(May, 3196)","(2, 6601)","(2.0, 7835)","(1.0, 4695)","(2, 3913)","(Returning_Visitor, 10551)","(False, 9273)","(False, 10303)"
bottom_value,"(Sept, 89)","(5, 6)","(9.0, 1)","(nan, 246)","(17, 1)","(Other, 85)","(Name:Zara, 189)","(nan, 147)"


### Page views and duration categories analysis:

Insights:

1. Administrative has negative values - this isn't possible as we're dealing with page views.
2. In all three cases the standard deviation is *higher* than the mean value - this suggests that total time spent in each page type bucket seems to vary a lot across different users.
3. 

Call to action:
1. Investigate rows with negative values - should these rows be remove or should be replace with 0? If time spent is 0, page views should be 0.
2. Find outliers in time spent and investigate whether they should be removed.
3. Consider feature engineering -> for each sample, divide duration with pageviews. Perhaps time spent on each page is relevant.

### Google Analytics categories analysis:

Insights:

1. BounceRates has negative values - this isn't possible as we're dealing with user behaviour.
2. BounceRates has values over 1 - this isn't possible with a percentage.
3. In all three cases the standard deviation is *higher* than the mean value - this suggests that behaviour seems to vary a lot across different users. 

Call to action:
1. Investigate rows with negative values - should these rows be remove or should be replace with 0?
2. Investigate rows with values over 1 - should they be removed?
3. Find outliers and investigate whether they should be removed.

### SpecialDay categories analysis:

Insights:

1. The standard deviation is *higher* than the mean value - this suggests that behaviour seems to vary a lot across different users.

Call to action:
1. Find outliers and investigate whether they should be removed.

In [126]:
# Contents of cells in the misc category df become truncated. We need to look into these.
# Extract misc df from list of dfs
misc_columns_df = summary_dfs_list[-1]

# Loop over columns in df and print unique values
print('Each column has the following unique values:', end=delimiter_between_prints)
for col in misc_columns_df.columns:

    print(col, misc_columns_df[col].loc['unique_values'])

print(end=delimiter_between_prints)

# Calculate ratio of converted rows to get a sense of how balanced our data is
print(f'Ratio of True values in Revenue column: {round(len(session_data_df[session_data_df["Revenue"] == True])/len(session_data_df), 4)}', end=delimiter_between_prints)

# We could see that we had at least one odd class in our mont column, 'Turc', lets count how many classes we have. We expect 12.
print(f"We have {len(misc_columns_df['Month'].loc['unique_values'])} unique classes in the Month column.")

Each column has the following unique values:

-----------

Month ['Feb' 'Mar' 'May' 'Turc' 'Oct' 'June' 'Jul' 'Aug' 'Nov' 'Sep' 'Sept'
 'Dec']
OperatingSystems [1 2 4 3 7 6 8 5]
Browser [nan  2.  1.  3.  4.  5.  6.  7. 10.  8.  9. 12. 13. 11.]
Region [ 1. nan  2.  3.  4.  9.  5.  6.  7.  8.]
TrafficType [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 18 19 16 17 20]
VisitorType ['Returning_Visitor' 'New_Visitor' 'Other']
Weekend ['False' 'True' 'Name:Zara']
Revenue [False nan True]


-----------

Ratio of True values in Revenue column: 0.1525

-----------

We have 12 unique classes in the Month column.


### Misc categories analysis:

Insights:

1. Month: while containing 12 classes as one might expect, January and April are missing. "Instead" we have "Sep" and "Sept" - are rows with these lables both belonging to September or could one of them actually be January or April but mislabled?
2. Month: Furthermore, we have a class called "Turc", could this be January or April that has been mislabled?
3. Browser: Contains nans
4. Region: Contains nans
5. VisitorType: "Other" is an odd class for visitor type - the documentation gives no clues to what it could be.
6. Weekend: Contains a class called "Name:Zara" which shouldn't appear in a column of boolean values.
7. Revenue: Imbalanced data set: only 15% of rows have True in the revenue column.

Call to action:
1. Month: Investigate Sep, Sept and Turc - are they January and April in disguise or erroneous samples?
2. Browser: Find nans and investigate removal or replacement
3. Region: Find nans and investigate removal or replacement
4. VisitorType: Find other and investigate keeping, removal or replacement
5. Weekend: Find "Name:Zara" and investigate removal or replacement
6. Revenue: Consider who to approach the issue of imbalance.

# We gotta stay positive . . .

In [127]:
# create a mask to filter out negative values in the Administrative column
admin_neg_values_filter = session_data_df['Administrative'] < 0

# apply the filter in our dataframe and save in a new dataframe
admin_neg_values_df = session_data_df[admin_neg_values_filter]

# have a quick glance at our new dataframe using the .info() method
admin_neg_values_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 123 entries, 143 to 12133
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Administrative           123 non-null    int64  
 1   Administrative_Duration  123 non-null    float64
 2   Informational            123 non-null    int64  
 3   Informational_Duration   123 non-null    float64
 4   ProductRelated           123 non-null    int64  
 5   ProductRelated_Duration  123 non-null    float64
 6   BounceRates              123 non-null    float64
 7   ExitRates                123 non-null    float64
 8   PageValues               123 non-null    float64
 9   SpecialDay               119 non-null    float64
 10  Month                    123 non-null    object 
 11  OperatingSystems         123 non-null    int64  
 12  Browser                  122 non-null    float64
 13  Region                   120 non-null    float64
 14  TrafficType              12

In [128]:
# In the rows with negative admin values, how many rows convert?
print(admin_neg_values_df['Revenue'].value_counts(), end=delimiter_between_prints)

admin_neg_values_df_target_true_count = len(admin_neg_values_df.loc[admin_neg_values_df['Revenue'] == True])
print(f'Ratio converted rows with neg value in admin vs total converted rows: {round(admin_neg_values_df_target_true_count/target_true_count_before_cleaning, 4)}', end=delimiter_between_prints)

admin_neg_values_df_row_count = len(admin_neg_values_df)
print(f'Ratio rows with neg value in admin vs total rows: {round(admin_neg_values_df_row_count/session_df_row_count, 4)}')

Revenue
False    106
True      16
Name: count, dtype: int64

-----------

Ratio converted rows with neg value in admin vs total converted rows: 0.0085

-----------

Ratio rows with neg value in admin vs total rows: 0.01


Insights:

1. 122 rows with negative values in Administrative column. <1% of total amount of rows.
2. Of these, 16 rows are sessions where convertion happened, which is <1% of all converted sessions.

Call to action:
1. If we assume the negative values became negative by accident, and make them positive, will they we representive of other rows?

In [129]:
# Let's look into the summaries of all the negative rows
summarise_df(admin_neg_values_df)

,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration
data_type,int64,float64,int64,float64,int64,float64
max_value,-1,1028.833333,9,1779.166667,338,13265.35595
min_value,-10,0.0,0,0.0,1,0.0
mean_value,-5.642276,68.211843,0.536585,63.774797,31.113821,1120.042206
standard_deviation,2.716465,150.812857,1.338538,261.488216,45.460161,1802.345859


,BounceRates,ExitRates,PageValues
data_type,float64,float64,float64
max_value,1.80952,0.2,120.587914
min_value,-0.38211,0.003226,0.0
mean_value,0.066005,0.047877,4.659112
standard_deviation,0.268121,0.055342,16.330311


,SpecialDay
data_type,float64
max_value,1.0
min_value,0.0
mean_value,0.112605
standard_deviation,0.261208


,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
data_type,object,int64,float64,float64,int64,object,object,object
unique_values,"[Feb, Mar, Turc, May, June, Oct, Aug, Jul, Nov...","[2, 1, 3, 8]","[2.0, nan, 1.0, 4.0, 6.0, 5.0, 10.0, 8.0]","[2.0, 1.0, nan, 4.0, 3.0, 7.0, 6.0, 9.0, 5.0, ...","[3, 1, 9, 8, 2, 10, 6, 7, 13, 4, 20, 5, 11]","[Returning_Visitor, New_Visitor]","[False, True, Name:Zara]","[False, True, nan]"
value_counts,"[(May, 34), (Nov, 30), (Mar, 27), (Dec, 13), (...","[(2, 63), (1, 35), (3, 24), (8, 1)]","[(2.0, 75), (1.0, 30), (4.0, 8), (5.0, 4), (8....","[(1.0, 52), (3.0, 22), (2.0, 12), (4.0, 11), (...","[(2, 37), (1, 23), (3, 21), (13, 9), (4, 8), (...","[(Returning_Visitor, 106), (New_Visitor, 17)]","[(False, 103), (True, 19), (Name:Zara, 1)]","[(False, 106), (True, 16), (nan, 1)]"
top_value,"(May, 34)","(2, 63)","(2.0, 75)","(1.0, 52)","(2, 37)","(Returning_Visitor, 106)","(False, 103)","(False, 106)"
bottom_value,"(Sep, 1)","(8, 1)","(10.0, 1)","(8.0, 2)","(5, 1)","(New_Visitor, 17)","(Name:Zara, 1)","(nan, 1)"


[                   Administrative Administrative_Duration Informational  \
 data_type                   int64                 float64         int64   
 max_value                      -1             1028.833333             9   
 min_value                     -10                     0.0             0   
 mean_value              -5.642276               68.211843      0.536585   
 standard_deviation       2.716465              150.812857      1.338538   
 
                    Informational_Duration ProductRelated  \
 data_type                         float64          int64   
 max_value                     1779.166667            338   
 min_value                             0.0              1   
 mean_value                      63.774797      31.113821   
 standard_deviation             261.488216      45.460161   
 
                    ProductRelated_Duration  
 data_type                          float64  
 max_value                      13265.35595  
 min_value                          

When we look at summaries of each column in the negative values dataframe, theres nothing that stands out. There is decent diversity of values in each column. While there are only 4 classes of OperatingSystems, it's not like there is only ONE. We need another metric of comparison.

Let's compare the mean value of duration per pageview for the rows where we have non-negative values for administrative with the mean value of duration per pageview for the rows with negative values for administrative.

Finding the mean of the non-negative rows.

In [130]:
# Create df for non-negative rows. We will call these positive for ease. But they include zeros.
admin_pos_values_df = session_data_df[~ admin_neg_values_filter]

# Create a new column that divides duration with page views
admin_pos_values_df['Duration per page view'] = admin_pos_values_df['Administrative_Duration'] / admin_pos_values_df['Administrative']

# Check if we got any infinities
admin_pos_values_df.loc[admin_pos_values_df['Duration per page view'] == np.inf]

C:\Users\SunnyVanderwall\AppData\Local\Temp\ipykernel_13988\1979418567.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  admin_pos_values_df['Duration per page view'] = admin_pos_values_df['Administrative_Duration'] / admin_pos_values_df['Administrative']


,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration,BounceRates,ExitRates,PageValues,SpecialDay,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue,Duration per page view
6434,0,971707.0,0,0.0,9,249.410000,0.022222,0.005128,0.0,0.0,Nov,1,2.0,1.0,4,Returning_Visitor,False,False,inf
8641,0,965480.0,0,0.0,1,0.000000,0.200000,0.200000,0.0,0.0,Dec,3,2.0,7.0,1,Returning_Visitor,False,False,inf
8979,0,984643.0,0,0.0,1,0.000000,0.200000,0.200000,0.0,0.0,Nov,1,1.0,1.0,1,Other,False,False,inf
9375,0,985147.0,0,0.0,42,2239.363889,0.020000,0.035455,0.0,0.0,Dec,3,2.0,7.0,2,Returning_Visitor,False,False,inf
9948,0,973917.0,0,0.0,3,0.000000,0.200000,0.200000,0.0,0.0,Nov,2,7.0,2.0,1,Returning_Visitor,False,False,inf


In [131]:
# Lets replace infinities with nan.
admin_pos_values_df.loc[admin_pos_values_df['Duration per page view'] == np.inf, 'Duration per page view'] = None

# Remove rows with Nans resulting from zero division issues. We only want to compare rows with a real ratio.
admin_pos_values_nans_removed_series = admin_pos_values_df['Duration per page view'].dropna()

admin_pos_values_mean = round(admin_pos_values_nans_removed_series.mean(), 4)
print(admin_pos_values_mean)

529.3628


In [132]:
# Save the ids of rows with negative admin values to we can use them later
rows_with_neg_admin_values_ids = admin_neg_values_df.loc[admin_neg_values_df['Administrative'] < 0, 'Administrative'].index

In [133]:
# Multiply all negative values by -1 to make them positive.
admin_neg_values_df['Administrative'] = admin_neg_values_df['Administrative'] * -1

# Check that it worked - this should result in an empty df
admin_neg_values_df[admin_neg_values_df['Administrative'] < 0]

C:\Users\SunnyVanderwall\AppData\Local\Temp\ipykernel_13988\662836193.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  admin_neg_values_df['Administrative'] = admin_neg_values_df['Administrative'] * -1


,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration,BounceRates,ExitRates,PageValues,SpecialDay,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue


In [134]:
admin_neg_values_df['Duration per page view'] = admin_neg_values_df['Administrative_Duration'] / admin_neg_values_df['Administrative']

admin_neg_values_mean = round(admin_neg_values_df['Duration per page view'].mean(), 4)

C:\Users\SunnyVanderwall\AppData\Local\Temp\ipykernel_13988\2779784053.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  admin_neg_values_df['Duration per page view'] = admin_neg_values_df['Administrative_Duration'] / admin_neg_values_df['Administrative']


In [135]:
print(f'The ratio of the mean of negative value rows vs positive value rows is: {round(admin_pos_values_mean / admin_neg_values_mean, 4)})')

The ratio of the mean of negative value rows vs positive value rows is: 34.977)


The mean value of the rows with negative values is 35 times smaller than the mean value of the rows with positve values. While this COULD be normal, its a large enough difference for us to want to remove the negative values from the dataset all together. They are small part of data set - its more pragmatic to remove them than replace their values.

In [136]:
session_data_df.drop(index=rows_with_neg_admin_values_ids, inplace=True)

summarise_df(session_data_df)

,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration
data_type,int64,float64,int64,float64,int64,float64
max_value,27,989493.0,24,2549.375,705,63973.52223
min_value,0,0.0,0,0.0,0,0.0
mean_value,2.317605,1281.883188,0.503236,34.177142,31.737691,1195.498951
standard_deviation,3.321804,34242.488,1.269502,138.988592,44.46733,1914.812818


,BounceRates,ExitRates,PageValues
data_type,float64,float64,float64
max_value,2.098952,0.2,361.763742
min_value,-0.49868,0.0,0.0
mean_value,0.04402,0.043024,5.901653
standard_deviation,0.239434,0.048524,18.589792


,SpecialDay
data_type,float64
max_value,1.0
min_value,0.0
mean_value,0.060754
standard_deviation,0.197884


,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
data_type,object,int64,float64,float64,int64,object,object,object
unique_values,"[Feb, Mar, May, Turc, Oct, June, Jul, Aug, Nov...","[1, 2, 4, 3, 7, 6, 8, 5]","[nan, 2.0, 1.0, 3.0, 4.0, 5.0, 6.0, 7.0, 10.0,...","[1.0, nan, 2.0, 3.0, 4.0, 9.0, 5.0, 6.0, 7.0, ...","[1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 9, 13, 14...","[Returning_Visitor, New_Visitor, Other]","[False, True, Name:Zara]","[False, nan, True]"
value_counts,"[(May, 3162), (Nov, 2968), (Mar, 1880), (Dec, ...","[(2, 6538), (1, 2550), (3, 2531), (4, 478), (8...","[(2.0, 7760), (1.0, 2394), (4.0, 717), (5.0, 4...","[(1.0, 4643), (3.0, 2323), (4.0, 1146), (2.0, ...","[(2, 3876), (1, 2428), (3, 2031), (4, 1061), (...","[(Returning_Visitor, 10445), (New_Visitor, 167...","[(False, 9170), (True, 2849), (Name:Zara, 188)]","[(False, 10197), (True, 1864), (nan, 146)]"
top_value,"(May, 3162)","(2, 6538)","(2.0, 7760)","(1.0, 4643)","(2, 3876)","(Returning_Visitor, 10445)","(False, 9170)","(False, 10197)"
bottom_value,"(Sept, 89)","(5, 6)","(9.0, 1)","(nan, 243)","(17, 1)","(Other, 85)","(Name:Zara, 188)","(nan, 146)"


[                   Administrative Administrative_Duration Informational  \
 data_type                   int64                 float64         int64   
 max_value                      27                989493.0            24   
 min_value                       0                     0.0             0   
 mean_value               2.317605             1281.883188      0.503236   
 standard_deviation       3.321804               34242.488      1.269502   
 
                    Informational_Duration ProductRelated  \
 data_type                         float64          int64   
 max_value                        2549.375            705   
 min_value                             0.0              0   
 mean_value                      34.177142      31.737691   
 standard_deviation             138.988592       44.46733   
 
                    ProductRelated_Duration  
 data_type                          float64  
 max_value                      63973.52223  
 min_value                          

# Nan of this makes sense . . . 

So it is kind of worrying to see that the Revenue column has nans. Since this is our target column that really aint good.
A target column contains the very values we are *trying to predict*. This makes having missing values more problematic than in feature columns.
The consensus within the ML community seems to be that one should generally avoid replacing missing values in the target column - instead it is better to remove them. This agrees with my own intuition since it seems weird to try and guess (even if it is educated) values that you are trying to predict.
I vote for removing these, buts lets crunch some numbers first.

In [137]:
revenue_is_nan_filter = session_data_df['Revenue'].isna()

revenue_is_nan_df = session_data_df[revenue_is_nan_filter]

display(revenue_is_nan_df)

print(f'There are {len(revenue_is_nan_df)} rows where revenue is a nan value. This amounts to {round(len(revenue_is_nan_df)/session_df_row_count, 4)} of the total rows.')

,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration,BounceRates,ExitRates,PageValues,SpecialDay,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
48,0,0.000000,0,0.0,11,800.833333,0.000000,0.003636,0.000000,0.0,Feb,3,2.0,1.0,3,Returning_Visitor,False,NaN
82,0,0.000000,0,0.0,6,39.500000,0.000000,0.020000,0.000000,0.0,Feb,2,2.0,1.0,3,Returning_Visitor,False,NaN
113,2,22.000000,0,0.0,25,436.500000,0.008000,0.024000,0.000000,0.2,Feb,2,4.0,1.0,1,Returning_Visitor,False,NaN
188,9,111.500000,1,48.5,49,1868.819697,0.000000,0.020709,1.706015,0.0,Mar,2,2.0,7.0,2,Returning_Visitor,False,NaN
200,1,32.000000,0,0.0,50,2867.000000,0.000000,0.004000,153.443248,0.0,Mar,2,2.0,7.0,8,Returning_Visitor,True,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12005,5,76.000000,0,0.0,66,1415.387500,0.003582,0.014262,9.829115,0.0,Dec,2,2.0,5.0,2,Returning_Visitor,False,NaN
12069,2,163.000000,0,0.0,136,4201.508333,0.000000,0.011265,0.000000,0.0,Dec,3,2.0,3.0,3,Returning_Visitor,False,NaN
12093,10,260.250000,0,0.0,31,1694.333333,0.017143,0.039048,15.108543,0.0,Nov,2,2.0,4.0,3,Returning_Visitor,False,NaN
12182,9,102.866667,3,339.5,29,693.254167,0.000000,0.014706,77.649145,0.0,Dec,2,10.0,1.0,2,Returning_Visitor,False,NaN


There are 146 rows where revenue is a nan value. This amounts to 0.0118 of the total rows.


The amount of rows where revenue is nan is only about 1% of the total amount of rows. This is a very small ratio so lets just remove them asap!

In [139]:
# Use .dropna() method with subset parameter to drop all rows with nans in the revenue column 
session_data_df.dropna(subset='Revenue', inplace=True)

Let's deal with the rest of the nans

In [140]:
columns_with_nans

{'SpecialDay': 123, 'Browser': 184, 'Region': 246, 'Revenue': 147}

In [142]:
for key in columns_with_nans:
        nans_df = session_data_df[session_data_df[key].isna()]
        print(f'Summary of rows with nans in column: {key}. Row count: {len(nans_df)}')
        summarise_df(nans_df)

Summary of rows with nans in column: SpecialDay. Row count: 118


,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration
data_type,int64,float64,int64,float64,int64,float64
max_value,16,970294.0,5,1037.15,501,21672.24425
min_value,0,0.0,0,0.0,1,0.0
mean_value,2.533898,8300.650148,0.542373,38.679708,29.508475,1169.7359
standard_deviation,3.46825,89315.71614,1.224005,125.131727,52.206812,2552.001745


,BounceRates,ExitRates,PageValues
data_type,float64,float64,float64
max_value,1.811388,0.2,261.491286
min_value,-0.147128,0.0,0.0
mean_value,0.059266,0.042636,8.730365
standard_deviation,0.253163,0.047099,29.340065


,SpecialDay
data_type,float64
max_value,NaN
min_value,NaN
mean_value,NaN
standard_deviation,NaN


,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
data_type,object,int64,float64,float64,int64,object,object,object
unique_values,"[Mar, May, Turc, Oct, Nov, Aug, Sep, Jul, Dec]","[1, 2, 3, 4]","[2.0, 5.0, 1.0, 10.0, 8.0, 6.0, 4.0, nan, 3.0]","[1.0, 4.0, 3.0, 6.0, 8.0, 2.0, 5.0, 9.0, 7.0, ...","[8, 1, 2, 3, 13, 4, 6, 5, 20, 10, 11]","[Returning_Visitor, New_Visitor]","[False, True, Name:Zara]","[False, True]"
value_counts,"[(May, 34), (Nov, 31), (Mar, 18), (Dec, 18), (...","[(2, 67), (1, 25), (3, 21), (4, 5)]","[(2.0, 78), (1.0, 23), (4.0, 8), (10.0, 2), (5...","[(1.0, 51), (3.0, 21), (4.0, 10), (2.0, 10), (...","[(2, 41), (1, 23), (3, 18), (4, 10), (13, 7), ...","[(Returning_Visitor, 104), (New_Visitor, 14)]","[(False, 95), (True, 22), (Name:Zara, 1)]","[(False, 98), (True, 20)]"
top_value,"(May, 34)","(2, 67)","(2.0, 78)","(1.0, 51)","(2, 41)","(Returning_Visitor, 104)","(False, 95)","(False, 98)"
bottom_value,"(Sep, 2)","(4, 5)","(3.0, 1)","(nan, 2)","(11, 1)","(New_Visitor, 14)","(Name:Zara, 1)","(True, 20)"


Summary of rows with nans in column: Browser. Row count: 181


,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration
data_type,int64,float64,int64,float64,int64,float64
max_value,22,1592.916667,9,1131.0,311,12500.35306
min_value,0,0.0,0,0.0,0,0.0
mean_value,2.491713,99.151543,0.508287,35.58113,28.353591,1161.904668
standard_deviation,3.832056,216.73279,1.241051,126.298272,38.555398,1721.850532


,BounceRates,ExitRates,PageValues
data_type,float64,float64,float64
max_value,2.098952,0.2,59.790143
min_value,-0.375127,0.00125,0.0
mean_value,0.075201,0.048768,3.725265
standard_deviation,0.303373,0.054304,10.19813


,SpecialDay
data_type,float64
max_value,0.8
min_value,0.0
mean_value,0.066667
standard_deviation,0.19605


,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
data_type,object,int64,float64,float64,int64,object,object,object
unique_values,"[Feb, Mar, May, Turc, Nov, June, Oct, Sep, Aug...","[1, 2, 3, 7, 4]",[nan],"[1.0, 3.0, 8.0, 4.0, 6.0, 2.0, 7.0, 5.0, nan, ...","[1, 3, 2, 13, 10, 4, 5, 6, 19, 20, 8, 11]","[Returning_Visitor, New_Visitor, Other]","[False, True, Name:Zara]","[False, True]"
value_counts,"[(May, 52), (Nov, 40), (Dec, 30), (Mar, 26), (...","[(2, 89), (3, 43), (1, 39), (4, 8), (7, 2)]","[(nan, 181)]","[(1.0, 67), (3.0, 35), (4.0, 15), (2.0, 14), (...","[(2, 63), (1, 37), (3, 26), (4, 16), (13, 11),...","[(Returning_Visitor, 149), (New_Visitor, 31), ...","[(False, 125), (True, 54), (Name:Zara, 2)]","[(False, 154), (True, 27)]"
top_value,"(May, 52)","(2, 89)","(nan, 181)","(1.0, 67)","(2, 63)","(Returning_Visitor, 149)","(False, 125)","(False, 154)"
bottom_value,"(Sept, 1)","(7, 2)","(nan, 181)","(nan, 4)","(11, 1)","(Other, 1)","(Name:Zara, 2)","(True, 27)"


Summary of rows with nans in column: Region. Row count: 241


,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration
data_type,int64,float64,int64,float64,int64,float64
max_value,22,1951.279141,7,1830.5,337,15454.43857
min_value,0,0.0,0,0.0,0,0.0
mean_value,2.431535,90.779929,0.53112,46.673686,29.394191,1161.878967
standard_deviation,3.721066,201.788447,1.172633,179.587833,38.601357,1840.580403


,BounceRates,ExitRates,PageValues
data_type,float64,float64,float64
max_value,1.96016,0.2,129.101374
min_value,-0.489344,0.0,0.0
mean_value,0.059086,0.048706,5.423886
standard_deviation,0.280409,0.053899,16.65181


,SpecialDay
data_type,float64
max_value,1.0
min_value,0.0
mean_value,0.059414
standard_deviation,0.193989


,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
data_type,object,int64,float64,float64,int64,object,object,object
unique_values,"[Feb, Mar, May, Turc, Oct, Aug, June, Jul, Nov...","[2, 4, 1, 3]","[2.0, 1.0, 6.0, 4.0, 8.0, 5.0, 10.0, 3.0, nan]",[nan],"[2, 3, 1, 6, 13, 9, 8, 4, 5, 11, 20, 10]","[Returning_Visitor, New_Visitor, Other]","[False, True, Name:Zara]","[False, True]"
value_counts,"[(May, 70), (Nov, 58), (Dec, 40), (Mar, 30), (...","[(2, 107), (1, 64), (3, 57), (4, 13)]","[(2.0, 148), (1.0, 64), (4.0, 8), (8.0, 4), (5...","[(nan, 241)]","[(2, 71), (1, 48), (3, 41), (4, 22), (13, 20),...","[(Returning_Visitor, 209), (New_Visitor, 29), ...","[(False, 179), (True, 60), (Name:Zara, 2)]","[(False, 209), (True, 32)]"
top_value,"(May, 70)","(2, 107)","(2.0, 148)","(nan, 241)","(2, 71)","(Returning_Visitor, 209)","(False, 179)","(False, 209)"
bottom_value,"(Sep, 3)","(4, 13)","(3.0, 3)","(nan, 241)","(9, 1)","(Other, 3)","(Name:Zara, 2)","(True, 32)"


Summary of rows with nans in column: Revenue. Row count: 0


IndexError: list index out of range

Nothing really stands out in the summary of columns per nan-category (Region, Browser, SpecialDay).

I wonder how many of these rows there are all together.


In [145]:
all_nans_combined_df = session_data_df[session_data_df.isna().any(axis=1)]

summarise_df(all_nans_combined_df)

all_nans_combined_row_count = len(all_nans_combined_df)
print(end=delimiter_between_prints)
print(f"There are {all_nans_combined_row_count} rows with nans. This is {round(all_nans_combined_row_count/session_df_row_count, 4)} of the total original row count", end=delimiter_between_prints)

converted_rows_in_all_nans_df = all_nans_combined_df[all_nans_combined_df['Revenue'] == True]
converted_rows_in_all_nans_row_count = len(converted_rows_in_all_nans_df)

print(f'There are {converted_rows_in_all_nans_row_count} rows with converted users in the nans subset. This is {round(converted_rows_in_all_nans_row_count/target_true_count_before_cleaning, 4)} of the total original count of converted rows.')

,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration
data_type,int64,float64,int64,float64,int64,float64
max_value,22,970294.0,9,1830.5,501,21672.24425
min_value,0,0.0,0,0.0,0,0.0
mean_value,2.465291,1911.410525,0.527205,41.269134,28.941839,1140.858889
standard_deviation,3.698356,42024.621062,1.209965,152.675788,41.508972,1890.381794


,BounceRates,ExitRates,PageValues
data_type,float64,float64,float64
max_value,2.098952,0.2,261.491286
min_value,-0.489344,0.0,0.0
mean_value,0.06256,0.047479,5.650305
standard_deviation,0.273193,0.052853,18.776759


,SpecialDay
data_type,float64
max_value,1.0
min_value,0.0
mean_value,0.060723
standard_deviation,0.192051


,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
data_type,object,int64,float64,float64,int64,object,object,object
unique_values,"[Feb, Mar, May, Turc, Nov, Oct, June, Aug, Sep...","[1, 2, 4, 3, 7]","[nan, 2.0, 1.0, 5.0, 6.0, 4.0, 10.0, 8.0, 3.0]","[1.0, nan, 3.0, 4.0, 8.0, 6.0, 2.0, 7.0, 5.0, ...","[1, 2, 3, 6, 8, 13, 10, 9, 4, 5, 11, 19, 20]","[Returning_Visitor, New_Visitor, Other]","[False, True, Name:Zara]","[False, True]"
value_counts,"[(May, 152), (Nov, 129), (Dec, 87), (Mar, 74),...","[(2, 259), (1, 128), (3, 119), (4, 25), (7, 2)]","[(2.0, 226), (nan, 181), (1.0, 87), (4.0, 15),...","[(nan, 241), (1.0, 118), (3.0, 56), (4.0, 25),...","[(2, 173), (1, 106), (3, 85), (4, 48), (13, 37...","[(Returning_Visitor, 455), (New_Visitor, 74), ...","[(False, 393), (True, 135), (Name:Zara, 5)]","[(False, 454), (True, 79)]"
top_value,"(May, 152)","(2, 259)","(2.0, 226)","(nan, 241)","(2, 173)","(Returning_Visitor, 455)","(False, 393)","(False, 454)"
bottom_value,"(Sept, 1)","(7, 2)","(3.0, 4)","(5.0, 7)","(19, 1)","(Other, 4)","(Name:Zara, 5)","(True, 79)"




-----------

There are 533 rows with nans. This is 0.0432 of the total original row count

-----------

There are 79 rows with converted users in the nans subset. This is 0.042 of the total original count of converted rows.


# We will have nan of this . . .

Just over 4% of rows have nans. Furthermore this subset contains just over 4% of all converted rows. Given that we have an imbalanaced dataset where only about 15% of of rows have our target value my intuition tells me it would not be good to remove these rows. A better apporach might be to replace the missing values.
In order to replace them we have a few alternatives.

1. Replace with mean, median or mode values of the column. This is a very inexact approach and we can quite easily do better than this.
2. Replace using KNN impute
3. Replace with 'Unknown'

We can actually create three dfs which we can then use when we train our models and see which approach works best:
1. A df with nans replaced with KNN impute
2. A df with nans replaced with 'Unknown'
3. A df with rows with nans removed

In [146]:
replace_nans_dict = {'SpecialDay' : 0, 'Browser' : 'Unknown', 'Region' : 'Unknown'}
session_data_nans_replaced_with_zero_and_unknown_df = session_data_df.fillna(replace_nans_dict)
any_olumns_with_nans_left = {key : value for key, value in session_data_nans_replaced_with_zero_and_unknown_df.isna().sum().to_dict().items() if value != 0}

if len(any_olumns_with_nans_left) == 0:
    print('All nans have been replaced')
else:
    print(any_olumns_with_nans_left)

summarise_df(session_data_nans_replaced_with_zero_and_unknown_df)

All nans have been replaced


,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration
data_type,int64,float64,int64,float64,int64,float64
max_value,27,989493.0,24,2549.375,705,63973.52223
min_value,0,0.0,0,0.0,0,0.0
mean_value,2.317552,1296.38498,0.50286,34.135346,31.7147,1194.762621
standard_deviation,3.323914,34448.875769,1.266783,139.093858,44.454774,1917.009821


,BounceRates,ExitRates,PageValues
data_type,float64,float64,float64
max_value,2.098952,0.2,361.763742
min_value,-0.49868,0.0,0.0
mean_value,0.044149,0.043088,5.889105
standard_deviation,0.239538,0.048609,18.542079


,SpecialDay
data_type,float64
max_value,1.0
min_value,0.0
mean_value,0.06036
standard_deviation,0.197358


,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
data_type,object,int64,object,object,int64,object,object,object
unique_values,"[Feb, Mar, May, Turc, Oct, June, Jul, Aug, Nov...","[1, 2, 4, 3, 7, 6, 8, 5]","[Unknown, 2.0, 1.0, 3.0, 4.0, 5.0, 6.0, 7.0, 1...","[1.0, Unknown, 2.0, 3.0, 4.0, 9.0, 5.0, 6.0, 7...","[1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12, 9, 13, 14...","[Returning_Visitor, New_Visitor, Other]","[False, True, Name:Zara]","[False, True]"
value_counts,"[(May, 3122), (Nov, 2932), (Mar, 1859), (Dec, ...","[(2, 6462), (1, 2520), (3, 2497), (4, 472), (8...","[(2.0, 7675), (1.0, 2364), (4.0, 704), (5.0, 4...","[(1.0, 4592), (3.0, 2290), (4.0, 1131), (2.0, ...","[(2, 3830), (1, 2398), (3, 1997), (4, 1050), (...","[(Returning_Visitor, 10319), (New_Visitor, 165...","[(False, 9064), (True, 2814), (Name:Zara, 183)]","[(False, 10197), (True, 1864)]"
top_value,"(May, 3122)","(2, 6462)","(2.0, 7675)","(1.0, 4592)","(2, 3830)","(Returning_Visitor, 10319)","(False, 9064)","(False, 10197)"
bottom_value,"(Sept, 87)","(5, 6)","(9.0, 1)","(Unknown, 241)","(17, 1)","(Other, 84)","(Name:Zara, 183)","(True, 1864)"


[                   Administrative Administrative_Duration Informational  \
 data_type                   int64                 float64         int64   
 max_value                      27                989493.0            24   
 min_value                       0                     0.0             0   
 mean_value               2.317552              1296.38498       0.50286   
 standard_deviation       3.323914            34448.875769      1.266783   
 
                    Informational_Duration ProductRelated  \
 data_type                         float64          int64   
 max_value                        2549.375            705   
 min_value                             0.0              0   
 mean_value                      34.135346        31.7147   
 standard_deviation             139.093858      44.454774   
 
                    ProductRelated_Duration  
 data_type                          float64  
 max_value                      63973.52223  
 min_value                          

In [147]:
# df with rows with nans dropped
session_data_nan_rows_dropped_df = session_data_df.dropna()

summarise_df(session_data_nan_rows_dropped_df)

,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration
data_type,int64,float64,int64,float64,int64,float64
max_value,27,989493.0,24,2549.375,705,63973.52223
min_value,0,0.0,0,0.0,0,0.0
mean_value,2.310722,1267.949118,0.501735,33.805513,31.842904,1197.254874
standard_deviation,3.305594,34059.822757,1.269387,138.432023,44.583784,1918.276308


,BounceRates,ExitRates,PageValues
data_type,float64,float64,float64
max_value,2.088658,0.2,361.763742
min_value,-0.49868,0.0,0.0
mean_value,0.043298,0.042885,5.900146
standard_deviation,0.237846,0.048397,18.531907


,SpecialDay
data_type,float64
max_value,1.0
min_value,0.0
mean_value,0.060965
standard_deviation,0.198466


,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
data_type,object,int64,float64,float64,int64,object,object,object
unique_values,"[Feb, Mar, May, Turc, Oct, June, Jul, Aug, Nov...","[3, 2, 1, 4, 7, 6, 8, 5]","[2.0, 3.0, 4.0, 1.0, 5.0, 6.0, 7.0, 10.0, 8.0,...","[2.0, 1.0, 3.0, 4.0, 9.0, 5.0, 6.0, 7.0, 8.0]","[4, 3, 5, 2, 1, 6, 7, 8, 10, 11, 12, 9, 13, 14...","[Returning_Visitor, New_Visitor, Other]","[False, True, Name:Zara]","[False, True]"
value_counts,"[(May, 2970), (Nov, 2803), (Mar, 1785), (Dec, ...","[(2, 6203), (1, 2392), (3, 2378), (4, 447), (8...","[(2.0, 7449), (1.0, 2277), (4.0, 689), (5.0, 4...","[(1.0, 4474), (3.0, 2234), (4.0, 1106), (2.0, ...","[(2, 3657), (1, 2292), (3, 1912), (4, 1002), (...","[(Returning_Visitor, 9864), (New_Visitor, 1584...","[(False, 8671), (True, 2679), (Name:Zara, 178)]","[(False, 9743), (True, 1785)]"
top_value,"(May, 2970)","(2, 6203)","(2.0, 7449)","(1.0, 4474)","(2, 3657)","(Returning_Visitor, 9864)","(False, 8671)","(False, 9743)"
bottom_value,"(Sept, 86)","(7, 5)","(9.0, 1)","(5.0, 303)","(17, 1)","(Other, 80)","(Name:Zara, 178)","(True, 1785)"


[                   Administrative Administrative_Duration Informational  \
 data_type                   int64                 float64         int64   
 max_value                      27                989493.0            24   
 min_value                       0                     0.0             0   
 mean_value               2.310722             1267.949118      0.501735   
 standard_deviation       3.305594            34059.822757      1.269387   
 
                    Informational_Duration ProductRelated  \
 data_type                         float64          int64   
 max_value                        2549.375            705   
 min_value                             0.0              0   
 mean_value                      33.805513      31.842904   
 standard_deviation             138.432023      44.583784   
 
                    ProductRelated_Duration  
 data_type                          float64  
 max_value                      63973.52223  
 min_value                          

First I would like to get to know each column a bit better. Lets start with the min, max, mean and std of each column.

In [ ]:
# Create four empty dicts according the category groups outlined earlier
# We will in the loops below add summarised values from each column to these dicts.
column_info_page_views_and_duration_summary = {}
column_info_google_analytics_summary = {}
column_info_time_categories_summary = {}
column_info_misc_categories_summary = {}

# Iterate over each column in the session DataFrame
for col in session_data_df.columns:

    # The Pandas API types module provides efficient functions for checking datatypes.
    # First lets check if the column contains numerical values.
    if pd.api.types.is_numeric_dtype(session_data_df[col]):

        # Next lets check which category grouping it belongs to.
        if col in column_info_page_views_and_duration_list:

            # For numerical regression data we want to know the max, min, mean values and the standard deviation
            column_info_page_views_and_duration_summary[col] = {
                                'data_type' : session_data_df[col].dtype,
                                'max_value' : session_data_df[col].max(),
                                'min_value' : session_data_df[col].min(),
                                'mean_value' : session_data_df[col].mean(),
                                'standard_deviation' : session_data_df[col].std()
                                }
            
        elif col in column_info_google_analytics_list:

            column_info_google_analytics_summary[col] = {
                                'data_type' : session_data_df[col].dtype,
                                'max_value' : session_data_df[col].max(),
                                'min_value' : session_data_df[col].min(),
                                'mean_value' : session_data_df[col].mean(),
                                'standard_deviation' : session_data_df[col].std()
                                }
            
        elif col in column_info_time_categories_list:

            column_info_time_categories_summary[col] = {
                                'data_type' : session_data_df[col].dtype,
                                'max_value' : session_data_df[col].max(),
                                'min_value' : session_data_df[col].min(),
                                'mean_value' : session_data_df[col].mean(),
                                'standard_deviation' : session_data_df[col].std()
                                }
        else:
            # For classification data, numerical or object, we want to know the unique values, their counts (including NaNs!) and top & bottom
            list_of_tuples_of_key_value_pairs_from_value_counts = [(key, int(value)) for key, value in session_data_df[col].value_counts(dropna=False).to_dict().items()]
            column_info_misc_categories_summary[col] = {
                                'data_type' : session_data_df[col].dtype,
                                'unique_values' : session_data_df[col].unique(),
                                'value_counts' : list_of_tuples_of_key_value_pairs_from_value_counts,
                                'top_value' : list_of_tuples_of_key_value_pairs_from_value_counts[0],
                                'bottom_value' : list_of_tuples_of_key_value_pairs_from_value_counts[-1]
                                }
    
    else:
        list_of_tuples_of_key_value_pairs_from_value_counts = [(key, int(value)) for key, value in session_data_df[col].value_counts(dropna=False).to_dict().items()]
        column_info_misc_categories_summary[col] = {
                                'data_type' : session_data_df[col].dtype,
                                'unique_values' : session_data_df[col].unique(),
                                'value_counts' : list_of_tuples_of_key_value_pairs_from_value_counts,
                                'top_value' : list_of_tuples_of_key_value_pairs_from_value_counts[0],
                                'bottom_value' : list_of_tuples_of_key_value_pairs_from_value_counts[-1]
                                }   

In [ ]:
page_views_and_duration_summary_df = pd.DataFrame(column_info_page_views_and_duration_summary)
page_views_and_duration_summary_df

Page views and duratioon categories analysis:

Insights:

1. Administrative has negative values - this isn't possible as we're dealing with page views.
2. In all three cases the standard deviation is *higher* than the mean value - this suggests that total time spent in each page type bucket seems to vary a lot across different users.
3. 

Call to action:
1. Investigate rows with negative values - should these rows be remove or should be replace with 0? If time spent is 0, page views should be 0.
2. Find outliers in time spent and investigate whether they should be removed.
3. Consider feature engineering -> for each sample, divide duration with pageviews. Perhaps time spent on each page is relevant.

In [ ]:
column_info_google_analytics_summary_df = pd.DataFrame(column_info_google_analytics_summary)
column_info_google_analytics_summary_df

Google Analytics categories analysis:

Insights:

1. BounceRates has negative values - this isn't possible as we're dealing with user behaviour.
2. BounceRates has values over 1 - this isn't possible with a percentage.
3. In all three cases the standard deviation is *higher* than the mean value - this suggests that behaviour seems to vary a lot across different users. 

Call to action:
1. Investigate rows with negative values - should these rows be remove or should be replace with 0?
2. Investigate rows with values over 1 - should they be removed?
3. Find outliers and investigate whether they should be removed.

In [ ]:
column_info_time_categories_summary_df = pd.DataFrame(column_info_time_categories_summary)
column_info_time_categories_summary_df

SpecialDay categories analysis:

Insights:

1. The standard deviation is *higher* than the mean value - this suggests that behaviour seems to vary a lot across different users.

Call to action:
1. Find outliers and investigate whether they should be removed.

In [ ]:
column_info_misc_categories_summary_df = pd.DataFrame(column_info_misc_categories_summary)

column_info_misc_categories_summary_df

The contents of several cells has been truncated, we need to see it all. Lets use a for loop and print to get a better picture.

In [ ]:
for col in column_info_misc_categories_summary_df.columns:

    print(col, column_info_misc_categories_summary_df[col].loc['unique_values'])

In [ ]:
print(f'Ratio of True values in Revenue column: {round(len(session_data_df[session_data_df["Revenue"] == True])/len(session_data_df), 4)}')

In [ ]:
len(column_info_misc_categories_summary_df['Month'].loc['unique_values'])

Misc categories analysis:

Insights:

1. Month: while containing 12 classes as one might expect, January and April are missing. "Instead" we have "Sep" and "Sept" - are rows with these lables both belonging to September or could one of them actually be January or April but mislabled?
2. Month: Furthermore, we have a class called "Turc", could this be January or April that has been mislabled?
3. Browser: Contains nans
4. Region: Contains nans
5. VisitorType: "Other" is an odd class for visitor type - the documentation gives no clues to what it could be.
6. Weekend: Contains a class called "Name:Zara" which shouldn't appear in a column of boolean values.
7. Revenue: Imbalanced data set: only 15% of rows have True in the revenue column.

Call to action:
1. Month: Investigate Sep, Sept and Turc - are they January and April in disguise or erroneous samples?
2. Browser: Find nans and investigate removal or replacement
3. Region: Find nans and investigate removal or replacement
4. VisitorType: Find other and investigate keeping, removal or replacement
5. Weekend: Find "Name:Zara" and investigate removal or replacement
6. Revenue: Consider who to approach the issue of imbalance.

_________________________________________________________________________________________________________________________________

Let's start working with our call-to-actions, starting with rows with negative values.

In [ ]:
# create a mask to filter out negative values in the Administrative column
admin_neg_values_filter = session_data_df['Administrative'] < 0

# apply the filter in our dataframe and save in a new dataframe
admin_neg_values_df = session_data_df[admin_neg_values_filter]

# have a quick glance at our new dataframe using the .info() method and by showing the dataframe in our notebook.
admin_neg_values_df.info()

In [ ]:
# In the rows with negative admin values, how many rows convert?
print(admin_neg_values_df['Revenue'].value_counts(), end=delimiter_between_prints)

admin_neg_values_df_target_true_count = len(admin_neg_values_df.loc[admin_neg_values_df['Revenue'] == True])
print(f'Ratio converted rows with neg value in admin vs total converted rows: {round(admin_neg_values_df_target_true_count/target_true_count_before_cleaning, 4)}', end=delimiter_between_prints)

admin_neg_values_df_row_count = len(admin_neg_values_df)
print(f'Ratio rows with neg value in admin vs total rows: {round(admin_neg_values_df_row_count/session_df_row_count, 4)}')

Insights:

1. 122 rows with negative values in Administrative column. <1% of total amount of rows.
2. Of these, 16 rows are sessions where convertion happened, which is <1% of all converted sessions.

Call to action:
1. If we assume the negative values became negative by accident, and make them positive, will they we representive of other rows?

In [ ]:
summarise_df(admin_neg_values_df)

When we look at summaries of each column in the negative values dataframe, theres nothing that stands out. There is decent diversity of values in each column. While there are only 4 classes of OperatingSystems, it's not like there is only ONE. We need another metric of comparison.

Let's compare the mean value of duration per pageview for the rows where we have non-negative values for administrative with the mean value of duration per pageview for the rows with negative values for administrative.

Finding the mean of the non-negative rows.

In [ ]:
# Create df for non-negative rows. We will call these positive for ease. But they include zeros.
admin_pos_values_df = session_data_df[~ admin_neg_values_filter]

# Create a new column that divides duration with page views
admin_pos_values_df['Duration per page view'] = admin_pos_values_df['Administrative_Duration'] / admin_pos_values_df['Administrative']

In [ ]:
# Check if we got any infinities
admin_pos_values_df.loc[admin_pos_values_df['Duration per page view'] == np.inf]

In [ ]:
# Lets replace infinities with nan.
admin_pos_values_df.loc[admin_pos_values_df['Duration per page view'] == np.inf, 'Duration per page view'] = None

In [ ]:
# Remove rows with Nans resulting from zero division issues. We only want to compare rows with a real ratio.
admin_pos_values_nans_removed_series = admin_pos_values_df['Duration per page view'].dropna()

admin_pos_values_nans_removed_series.mean()

In [ ]:
# Save the ids of rows with negative admin values to we can use them later
rows_with_neg_admin_values_ids = admin_neg_values_df.loc[admin_neg_values_df['Administrative'] < 0, 'Administrative'].index


In [ ]:
# Multiply all negative values by -1 to make them positive.
admin_neg_values_df['Administrative'] = admin_neg_values_df['Administrative'] * -1

# Check that it worked - this should result in an empty df
admin_neg_values_df[admin_neg_values_df['Administrative'] < 0]

In [ ]:
admin_neg_values_df['Duration per page view'] = admin_neg_values_df['Administrative_Duration'] / admin_neg_values_df['Administrative']

admin_neg_values_df['Duration per page view'].mean(),

In [ ]:
admin_pos_values_nans_removed_series.mean() / admin_neg_values_df['Duration per page view'].mean()

The mean value of the rows with negative values is 35 times smaller than the mean value of the rows with positve values. While this COULD be normal, its a large enough difference for us to want to remove the negative values from the dataset all together. They are small part of data set - its more pragmatic to remove them than replace their values.

In [ ]:
session_data_df.drop(index=rows_with_neg_admin_values_ids, inplace=True)

summarise_df(session_data_df)

**Page views > 0 *and* duration = 0**

While it intuitively feels like the page views with 0 duration are errors, it could be that there are technical quirks that cause this to happen. And that these are legitimate samples.

It quite hard to make judgemnt calls without a domain expert or the people who gave us the data. However my research on the topic suggests that these could be legit user sessions where:
- The user entered a page without interacting with it and then bounced. This generates a zero value for duration in Google analytics
- The website analytics isn't configured in a way that captures those visits correctly

Lets checkout those zero duration sessions further, by category:

In [ ]:
admin_duration_zero_filter = session_data_df['Administrative_Duration'] == 0
info_duration_zero_filter = session_data_df['Informational_Duration'] == 0
product_duration_zero_filter = session_data_df['ProductRelated_Duration'] == 0

admin_duration_zero_df = session_data_df[admin_duration_zero_filter]
info_duration_zero_df = session_data_df[info_duration_zero_filter]
product_duration_zero_df = session_data_df[product_duration_zero_filter]

In [ ]:
summarise_df(admin_duration_zero_df)

In [ ]:
admin_duration_zero_df['Administrative'].value_counts()

Insights:

1. Most zero duration rows also have 0 page views. This makes intuitive sense and is fine.
2. 128 rows have 1 page view - which could be explained by the fact that a user maybe have entered a page, not interacting in it and then bounced which would result in a zero in duration
3. A very small number (4) of rows have *2* page views, but still with 0 duration. This seems intuitively odd but could be down to some technical reason that I am not aware of. Given the small amount I did not research it extensively at this time.

Call to action:

Nothing. Keep rows as is.

In [ ]:
summarise_df(info_duration_zero_df)

In [ ]:
info_duration_zero_df['Informational'].value_counts()

Insights:

1. Most zero duration rows also have 0 page views. This makes intuitive sense and is fine.
2. 211 rows have 1 page view - which could be explained by the fact that a user maybe have entered a page, not interacting in it and then bounced which would result in a zero in duration
3. A very small number (9) of rows have *2* page views, but still with 0 duration. This seems intuitively odd but could be down to some technical reason that I am not aware of. Given the small amount I did not research it extensively at this time.

Call to action:

Nothing. Keep rows as is.

In [ ]:
summarise_df(product_duration_zero_df)

In [ ]:
product_duration_zero_df['ProductRelated'].value_counts()

Insights:

This is a more complicated picture: While the majority of zero duration session in this category is 0 or 1 sessions - there are over 100 sessions with 2 or more page views with zero duration. This is harder to understand intuitively. However since the number is still quite low, and this could be easily explainable behaviour for a domain expert, I will leave it in for now.

I will just make a quick summary of all columns for this particular subset and see if I can, at a glance, see anything that sticks out.

In [ ]:
product_duration_zero_product_above_one_df = product_duration_zero_df[product_duration_zero_df['ProductRelated'] > 1]

summarise_df(product_duration_zero_product_above_one_df)

In [ ]:
product_duration_zero_product_one_and_zero_df = product_duration_zero_df[product_duration_zero_df['ProductRelated'] <= 1]

summarise_df(product_duration_zero_product_one_and_zero_df)

Nothing sticks out. Lets keep the rows and keep going. GAMBARU!